code/code/misc/info.cc

```diff
@@ -5385,642 +5385,291 @@ void TBeing::describeTrapDamType(const TTrap *obj, int) const
+  sort(skillsSorted.begin(), skillsSorted.end(), skillSorter());
+
+  auto getSpellCategory = [](spellInfo *spell) {
+    bool isViolent = spell->targets & TAR_VIOLENT;
+    bool isAOE = spell->targets & TAR_AREA;
+    bool isTargeted = spell->targets & TAR_CHAR_ROOM;
+
+    return isViolent
+      ? isAOE
+        ? NONTARGETED_OFFENSIVE
+        : TARGETED_OFFENSIVE
+      : isTargeted
+        ? TARGETED_UTILITY
+        : NONTARGETED_UTILITY;
+  };
 
-  for (type = 0; type < 4;++type) {
-    if (!types[type])
-      continue;
+  for (auto &skill : skillsSorted) {
+    spellNumT skillNumber = skill.theSkill;
+    spellInfo *spell = discArray[skillNumber];
 
-    if (*buffer)
-      strcat(buffer, "\n\r");
+    spellInfoByType[spell->typ][getSpellCategory(spell)].push_back(skillNumber);
+  }
 
-    switch (type) {
-      case 0:
-        strcat(buffer, "Targeted offensive rituals:\n\r");
-        break;
-      case 1:
-        strcat(buffer, "Non-targeted offensive rituals:\n\r");
-        break;
-      case 2:
-        strcat(buffer, "Targeted utility rituals:\n\r");
-        break;
-      case 3:
-        strcat(buffer, "Non-targeted utility rituals:\n\r");
-        break;
-    }
+  auto getComponentString = [this](spellNumT spellNumber) {
+    auto comp = std::find_if(CompInfo.begin(), CompInfo.end(),
+                             [spellNumber](compInfo &ci) { return ci.spell_num == spellNumber; });
 
-    for (j = 0; j < skillSortVec.size(); j++) {
-      i = skillSortVec[j].theSkill;
-      das = getDisciplineNumber(i, FALSE);
-      if (das == DISC_NONE) {
-        vlogf(LOG_BUG, format("Bad disc for skill %d in doRituals") %  i);
-        continue;
-      }
-      cd = getDiscipline(das);
-      
-      // getLearnedness is -99 for an unlearned skill, make it seem like a 0
-      int tmp_var = ((!cd || cd->getLearnedness() <= 0) ? 0 : cd->getLearnedness());
-      tmp_var = min((int) MAX_DISC_LEARNEDNESS, tmp_var);
-
-      switch (type) {
-        case 0: // single target offensive
-          if (!(discArray[i]->targets & TAR_VIOLENT) ||
-              (discArray[i]->targets & TAR_AREA))
-            continue;
-          break;
-        case 1: // area offensive
-          if (!(discArray[i]->targets & TAR_VIOLENT) ||
-              !(discArray[i]->targets & TAR_AREA))
-            continue;
-          break;
-        case 2: // targeted utility
-          if ((discArray[i]->targets & TAR_VIOLENT) ||
-              !(discArray[i]->targets & TAR_CHAR_ROOM))
-            continue;
-          break;
-        case 3: // non-targeted utility
-          if ((discArray[i]->targets &  TAR_VIOLENT) ||
-              (discArray[i]->targets & TAR_CHAR_ROOM))
-            continue;
-          break;
-      }
+    if (comp != CompInfo.end() && comp->comp_num > 0) {
+      std::vector<TThing *> possibleComponentLocations = {
+          heldInPrimHand(),       heldInSecHand(),      this,
```

Comment:
Agreed. 

I've spent some time looking into this and so far it doesn't appear to be possible with clang-format through version 12, which is the highest available through the official repos in Ubuntu Focal (which uses clang-format-10 by default). 

There are some newer rules that become available with clang-format 13 and 14 that might work to format this the way we want, but that'll require upgrading the Docker images to the newest Ubuntu LTS (22.04, currently, I believe) or manually installing the newest clang-format in Focal, perhaps?

I can format this part manually if you'd like, but it'll get reverted if we do an auto-format pass before upgrading clang-format.

Reply:
We're not married to Ubuntu Focal. We used to use Focal because it was the newest base image in Travis CI.
In fact, we're not tied to Ubuntu at all. Might want to look into Alpine, it's normally the smallest one.